### DISCUSSION TITLE
Complex Jooq transactions (in spring boot 3)?

### DISCUSSION THREAD

**Discussion by 0248991** [jOOQ/jOOQ#14092] [Q&A] [answered]
Hello,
I read many blogs, tutorials, and watched quite a few videos about how to do transactions with JOOQ, but I haven't found a simple and satisfactory answer, so I would like to ask for some more details and clarifications.

The documentation page https://www.jooq.org/doc/latest/manual/sql-execution/transaction-management/ gives an example how to create a transaction:

```
create.transaction((Configuration trx) -> {

    // Wrap configuration in a new DSLContext:
    trx.dsl().insertInto(...);
    trx.dsl().insertInto(...);
    
    // Or, reuse the new DSLContext within the transaction scope:
    DSLContext ctx = trx.dsl();
    ctx.insertInto(...);
    ctx.insertInto(...);
    
    // ... but avoid using the scope from outside the transaction:
    create.insertInto(...);
    create.insertInto(...);
});
```

First, I would like to ask what are `DSLContext` and `Configuration` classes.
Is `Configuration` basically a wrapper class around values required to connect to a database (username, password, URL,...)?
Is `DSLContext` basically a threadsafe builder (an entrypoint) for creating queries, which uses `Configuration` in the background to submit them to a database?

If so, does this mean the `DSLContext` bean spring boot auto-configures (and is able to autowire anywhere in an application) is not really ready for a propagation of a single transaction through a complex business code (including OOP patterns)?

E.g.:
```
class UserService(
	val dslContext: DSLContext
) {

	fun storeTwoUsers() {
		dslContext.transaction(cfg -> {
			val context = cfg.dsl();
			
			//how to use the transaction context?
			//should it be passed as an argument?
			//passing the variable through a whole system seems ridiculous
			
			//I want to store two users in a transaction
			storeSingleUser()
			storeSingleUser()
		
		});
	}

	fun storeSingleUser() {
		dslContext.insert(USER, USER.NAME)
		.values("single user")
		.execute()
	}
}
```

Passing the transaction context around seems like a bad idea.
But the solution I found is to use spring's `@transactional` annotation like so:
```
class UserService(
	val dslContext: DSLContext
) {

	@Transactional
	fun storeTwoUsers() {
		storeSingleUser()
		storeSingleUser()
	}

	fun storeSingleUser() {
		dslContext.insert(USER, USER.NAME)
		.values("single user")
		.execute()
	}
}
```

Here I would like to ask:
- is this real and does it work?
- is any additional configuration needed? (I read something about configuring a transaction manager or something...)
- are there plans to make everything auto-configured and work out of the box?
- what about support for the upcoming spring boot 3?

Thank you :)

**Comment by lukaseder** [comment, marked as answer]
Thanks a lot for your question.

> First, I would like to ask what are `DSLContext` and `Configuration` classes.

You'll find some information about code examples here:

<img width="785" alt="image" src="https://user-images.githubusercontent.com/734593/195538504-3e30c491-0b56-4f9f-8a9d-eb6476ce6d93.png">

Also here:
https://blog.jooq.org/a-brief-overview-over-the-most-common-jooq-types/

And the relevant Javadoc. I think your questions are already answered there.

> Is `Configuration` basically a wrapper class around values required to connect to a database (username, password, URL,...)?

Yes

> Is `DSLContext` basically a threadsafe builder (an entrypoint) for creating queries, which uses `Configuration` in the background to submit them to a database?

Please refer to the Javadoc, or the manual, it explains these ideas.

> If so, does this mean the `DSLContext` bean spring boot auto-configures (and is able to autowire anywhere in an application) is not really ready for a propagation of a single transaction through a complex business code (including OOP patterns)?

I don't know what you mean by "not really ready". But your example is wrong. If you want to use the jOOQ transaction API, you have to pass around the transactional `Configuration`. You can't use the injected `Configuration`, which isn't transactional. jOOQ's `DefaultTransactionProvider` isn't thread-bound. There's a `ThreadLocalTransactionProvider`, but you weren't using that.

> But the solution I found is to use spring's `@transactional` annotation like so:

Yes, why not just do that? You don't *have to* use jOOQ's transaction API. In fact, if you're using Spring, then why not just use Spring for transaction management?

> * is this real and does it work?

I don't know what you mean by "is this real". Your first code example doesn't work for the reasons I mentioned above.

> * is any additional configuration needed? (I read something about configuring a transaction manager or something...)

No, you don't need additional configuration. Spring Boot auto configures everything for you. What you read about transaction managers is when you use vanilla Spring. Then you do the work Spring Boot does manually.

> * are there plans to make everything auto-configured and work out of the box?

It already works. I guess all of this is more of a question directed at Spring Boot folks?

> * what about support for the upcoming spring boot 3?

We'll look into it when it's released. Any questions you have about Spring Boot 3 and jOOQ are best directed at Spring Boot folks. Do you have *specific* questions? Or just generic doubts? I mean, I don't think anything jOOQ related will change fundamentally in Spring Boot 3. Why would it?

**Comment by 0248991** [comment]
Hello,
I am unsure about the levels of abstraction spring boot uses to call a database.

The design is that somewhere on the bottom, there runs a Hikari pool that uses a specific JDBC driver.
Then the pool is wrapped by the `DataSource` class (and then by the `EntityManager` class?).
Does this sound about right?

If so, does spring boot's `@Transactional` annotation directly use the `DataSource` or the `EntityManager` to handle (start, commit, rollback) transactions?
Is that why it works with both spring data and JOOQ implicitly? Because it's implemented on a lower level?

Also, JOOQ always flushes queries executed by a code, right?
So stuff like query reordering doesn't happen, that's just JPA/Hibernate thing, right?

Thank you very much for the effort you put into the answer :)

**Reply by lukaseder** [reply]
jOOQ is just type safe JDBC with sophisticated mapping. That vision should answer all of your questions, I think?

**Reply by 0248991** [reply]
I guess to, thank you :)
